cli/azd/internal/telemetry/uploader.go

```diff
@@ -0,0 +1,193 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	appinsightsexporter "github.com/azure/azure-dev/cli/azd/internal/telemetry/appinsights-exporter"
+	"github.com/benbjohnson/clock"
+	"github.com/sethvargo/go-retry"
+)
+
+const (
+	maxRetryCount       = 3
+	maxStorageFailCount = 5
+)
+
+var (
+	// Optimistic first retry delay
+	firstTransmitRetryDelay   = time.Duration(100) * time.Millisecond
+	defaultTransmitRetryDelay = time.Duration(2) * time.Second
+	storageQueueRetryDelay    = time.Duration(300) * time.Millisecond
+
+	defaultThrottleDuration = time.Duration(5) * time.Second
+)
+
+type Uploader interface {
+	Upload(ctx context.Context, result chan (error))
+}
+
+type TelemetryUploader struct {
+	transmitter    appinsightsexporter.Transmitter
+	telemetryQueue Queue
+	clock          clock.Clock
+
+	isDebugMode bool
+}
+
+func NewUploader(telemetryQueue Queue, transmitter appinsightsexporter.Transmitter, clock clock.Clock, isDebugMode bool) *TelemetryUploader {
+	return &TelemetryUploader{
+		transmitter:    transmitter,
+		telemetryQueue: telemetryQueue,
+		clock:          clock,
+		isDebugMode:    isDebugMode,
+	}
+}
+
+// Uploads all items that are currently in the telemetry queue.
+// This function returns when no items remain in the queue.
+// An error is only returned if there is a fatal, persistent error with reading the queue.
+func (u *TelemetryUploader) Upload(ctx context.Context, result chan (error)) {
+	for {
+		select {
+		case <-ctx.Done():
+			result <- ctx.Err()
+			return
+		default:
+			done, err := u.uploadNextItem(ctx)
+
+			if done {
+				result <- err
+				return
+			}
+		}
+	}
+}
+
+func (u *TelemetryUploader) uploadNextItem(ctx context.Context) (bool, error) {
+	item, err := u.reliablePeek(ctx)
+
+	if err != nil {
+		return true, fmt.Errorf("failed to peek after %d attempts: %w", maxStorageFailCount, err)
+	}
+
+	if item == nil {
+		return true, nil
+	}
+
+	u.transmit(ctx, item)
+	err = u.reliableRemove(ctx, item)
+
+	if err != nil {
+		return true, fmt.Errorf("failed to remove after %d attempts: %w", maxStorageFailCount, err)
+	}
+
+	return false, nil
+}
+
+func (u *TelemetryUploader) reliablePeek(ctx context.Context) (*StoredItem, error) {
+	var item *StoredItem
+	err := retry.Do(ctx, retry.WithMaxRetries(maxStorageFailCount, retry.NewConstant(storageQueueRetryDelay)), func(ctx context.Context) error {
+		peekItem, err := u.telemetryQueue.Peek()
+
+		if err != nil {
+			return retry.RetryableError(err)
+		}
+
+		item = peekItem
+		return nil
+	})
+
+	if err != nil && ctx.Err() != nil {
+		// Attempt fallback - remove and retry Peek
+		err = u.reliableRemove(ctx, item)
+
+		if err != nil {
+			item, err = u.telemetryQueue.Peek()
```

Comment:
should this recur instead of calling `Peek` directly so that we get the retry behavior as above?